### PR TITLE
openslide: new recipe

### PR DESCRIPTION
**Pull request by valgur**
Adds OpenSlide: https://github.com/openslide/openslide

An optional dependency for libvips (https://www.libvips.org/).

[Packaging status](https://repology.org/project/openslide/versions)

### CODE REVIEW

File: recipes/openslide/all/test_package/test_package.c [hunk 1]
```diff
@@ -0,0 +1,6 @@
+#include <openslide.h>
+
+int main() {
+    openslide_detect_vendor("");
```
Comment (bgilbert) [inline recipes/openslide/all/test_package/test_package.c:4]: This doesn't really test anything.  You'd need to do `openslide_open("")` instead, with the `OPENSLIDE_DEBUG` environment variable set to `synthetic`, and you'd need to check whether the open succeeded.  `meson test` will do all of this for you, and I'd recommend using that rather than hand-rolling your own test.  If you want to avoid running tests at build time, you could run `OPENSLIDE_DEBUG=synthetic slidetool prop list ""` instead.

File: recipes/openslide/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+    def layout(self):
+        basic_layout(self, src_folder="src")
+
+    def requirements(self):
+        self.requires("cairo/1.18.0")
+        self.requires("gdk-pixbuf/2.42.10")
+        self.requires("glib/2.78.1", force=True)  # FIXME
+        self.requires("libdicom/1.0.5")
+        self.requires("libjpeg/9e")
```
Comment (bgilbert) [inline recipes/openslide/all/conanfile.py:51]: libjpeg-turbo is faster and widely used.  I'd recommend using it instead.

**Review by bgilbert — COMMENTED** [review]
Thanks for doing this!

File: recipes/openslide/all/test_package/test_package.c (hunk 1 above)
Comment (valgur) [reply recipes/openslide/all/test_package/test_package.c:4]: > This doesn't really test anything.

As intended. :smile: `test_package` exists only to confirm that compilation, linking and dynamic linking during execution actually work.

As for unit tests - CCI unfortunately does not allow running actual tests as part of the build since it would be too taxing for the CI given the quite large build matrix.

File: recipes/openslide/all/conanfile.py (hunk 2 above)
Comment (valgur) [reply recipes/openslide/all/conanfile.py:51]: Would love to, but due to dependency conflicts CCI has to stick to one variant of the library and has settled on `libjpeg` as the default. I personally hope that this is something that could be changed after the Conan 2.0 migration effort is complete.

I added an option for the choice of `libjpeg` variant at least.

**Comment by valgur** [comment]
> Thanks for doing this!

@bgilbert No problem, thanks for dropping by and adding your review!

File: recipes/openslide/all/conanfile.py [hunk 3]
```diff
@@ -58,6 +60,10 @@ def requirements(self):
         self.requires("sqlite3/3.44.0")
         self.requires("zlib/[>=1.2.11 <2]")
 
+    def validate(self):
+        if is_msvc(self):
+            raise ConanInvalidConfiguration(f"OpenSlide requires GNU C++ extensions support and is not compatible with MSVC")
```
Comment (bgilbert) [inline recipes/openslide/all/conanfile.py:72]: Nit: C, not C++.

File: recipes/openslide/all/test_package/test_package.c (hunk 1 above)
Comment (bgilbert) [reply recipes/openslide/all/test_package/test_package.c]: Ah, got it.  If the rules don't absolutely forbid it, you might still consider running the built-in smoke test though.  Its purpose is exactly to check that each of the dependencies is linked correctly and can read a built-in 16x16 test image.  `meson test` runs _only_ that test right now (most of our tests require a lot of slide data), but if you want to be future-proof, you could avoid `meson test` and directly open the zero-length filename as mentioned above.  On my machine, invoking it via `slidetool prop list ""` takes less than 30 ms of CPU time.

File: recipes/openslide/all/conanfile.py (hunk 3 above)
Comment (valgur) [reply recipes/openslide/all/conanfile.py]: Thanks!

File: recipes/openslide/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+from conan.errors import ConanInvalidConfiguration
+from conan.tools.apple import fix_apple_shared_install_name
+from conan.tools.env import VirtualBuildEnv
+from conan.tools.files import copy, get, rm, rmdir
+from conan.tools.gnu import PkgConfigDeps
+from conan.tools.layout import basic_layout
+from conan.tools.meson import Meson, MesonToolchain
+from conan.tools.microsoft import is_msvc
+
+required_conan_version = ">=1.53.0"
```
Comment (RubenRBS) [inline recipes/openslide/all/conanfile.py:13]: Will need >1.60 <2 || >=2.0.5 fopr this one as it's using the `<host_version>` syntax

File: recipes/openslide/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+        self.settings.rm_safe("compiler.cppstd")
+        self.settings.rm_safe("compiler.libcxx")
+
+    def layout(self):
+        basic_layout(self, src_folder="src")
+
+    def requirements(self):
+        self.requires("cairo/1.18.0")
+        self.requires("gdk-pixbuf/2.42.10")
+        self.requires("glib/2.78.1", force=True)  # FIXME
```
Comment (RubenRBS) [inline recipes/openslide/all/conanfile.py:55]: Why are these needed?

**Review by RubenRBS — COMMENTED** [review]
Thanks! Rebuilding the missing dependency now

File: recipes/openslide/all/conanfile.py (hunk 5 above)
Comment (valgur) [reply recipes/openslide/all/conanfile.py:55]: It's blocked by #21137.

**Comment by valgur** [comment]
Blocked by 
- #23143

**Review by AbrilRBS — APPROVED** [review]
Thanks a lot for your efforts and patience, this looks great to go now :)

File: recipes/openslide/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+    def package_info(self):
+        self.cpp_info.libs = ["openslide"]
+        self.cpp_info.includedirs.append(os.path.join("include", "openslide"))
+
+        if self.settings.os == "Macos":
+            self.cpp_info.frameworks = ["CoreText"]
+
+        self.cpp_info.requires = [
+            "cairo::cairo_",
```
Comment (AbrilRBS) [inline recipes/openslide/all/conanfile.py:120]: (Thanks @czoido for spotting this!) Why is the virtual main component used here instead of the whole `cairo::cairo` global one?

Comment (valgur) [reply recipes/openslide/all/conanfile.py:120]: That's done intentionally to match the component used by the project. It also avoids overlinking and unnecessary transitive dependencies, somewhat.

**Review by AbrilRBS — APPROVED** [review]

**Review by danimtb — APPROVED** [review]
